petastorm/pytorch.py

```diff
@@ -348,7 +365,38 @@ def _instantiate_noop_buffer():
 
         return shuffling_buffer, other_shuffling_buffer, instantiate_buffer_fn
 
+    def _worker(self):
+        for b in self._iter_impl_worker():
+            if self._finished_event.is_set():
+                break
+            self._async_batch_queue.put(b)
+        else:
+            self._async_batch_queue.put(None)
+
     def _iter_impl(self):
+        if self._async_batch_queue:
+            # Get shuffled batches from async shuffle thread
+            if not self._started:
+                # If reuse dataloader in a new pass, async shuffle thread has to be relaunched.
+                # It is finished in previous pass when reaching end of data.
+                self._shuffle_thread = Thread(target=self._worker)
+                self._shuffle_thread.daemon = True
```

Comment:
Is there a reason for the thread being daemonic? My concern is that random exceptions might show up on shutdown if we don't properly shutdown threads.

Reply:
I see. Make sense to me.